File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func removeNthFromEnd(head *ListNode, n int) *ListNode {
    dummyNode := &ListNode{}

    dummyNode.Next = head

    ahead, behind := head, dummyNode

    for i := 0; i < n; i++ {
        ahead = ahead.Next
    }

    for ahead != nil {
        ahead = ahead.Next
        behind = behind.Next
    }

    behind.Next = behind.Next.Next

    return dummyNode.Next
}

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummyNode = ListNode()
        dummyNode.next = head
        ahead, behind = head, dummyNode

        for _ in range(n):
            ahead = ahead.next
        
        while ahead:
            ahead = ahead.next
            behind = behind.next
        
        behind.next = behind.next.next

        return dummyNode.next
        
